xdsl/transforms/stencil_to_csl_stencil.py

```diff
@@ -1,23 +1,146 @@
 from collections.abc import Sequence
-
-from attr import dataclass
+from dataclasses import dataclass
+from math import prod
 
 from xdsl.context import MLContext
-from xdsl.dialects import memref, stencil, tensor
-from xdsl.dialects.builtin import ModuleOp, TensorType
+from xdsl.dialects import arith, memref, stencil, tensor
+from xdsl.dialects.builtin import (
+    DenseArrayBase,
+    IndexType,
+    IntegerAttr,
+    IntegerType,
+    ModuleOp,
+    TensorType,
+)
 from xdsl.dialects.csl import csl_stencil
 from xdsl.dialects.experimental import dmp
-from xdsl.ir import Attribute, OpResult
+from xdsl.ir import Attribute, Block, BlockArgument, Operation, OpResult, Region
+from xdsl.irdl import Operand
 from xdsl.passes import ModulePass
 from xdsl.pattern_rewriter import (
+    GreedyRewritePatternApplier,
     PatternRewriter,
     PatternRewriteWalker,
     RewritePattern,
     op_type_rewrite_pattern,
 )
+from xdsl.rewriter import InsertPoint
+from xdsl.transforms.experimental.stencil_tensorize_z_dimension import (
+    BackpropagateStencilShapes,
+)
 from xdsl.utils.hints import isa
 
 
+def get_stencil_accessed_symbols(op: Operand) -> set[Operand]:
+    """
+    Returns the symbols of all stencil accessess by op and all its dependencies.
+    """
+    res: set[Operand] = set()
+    frontier: set[Operand] = set((op,))
+    next: set[Operand] = set()
+    while len(frontier) > 0:
+        for o in frontier:
+            if isinstance(o, OpResult):
+                if isinstance(o.op, csl_stencil.AccessOp):
+                    res.add(o.op.op)
+                else:
+                    next.update(o.op.operands)
+        frontier = next
+        next = set()
+
+    return res
+
+
+@dataclass(frozen=True)
+class RestructureSymmetricReductionPattern(RewritePattern):
+    """
+    Consume data first where that data comes from stencil accesses to `buf`.
+
+    Identifies a pattern of 2 connected binary ops with 3 args, e.g. of the form `(a+b)+c` with different ops and
+    bracketings supported, and attempts to re-structure the order of computation.
+
+    Being in principle similarly to constant folding, the difference is that args are *not* required to be stencil
+    accesses, but could have further compute applied before being passed to the reduction function.
+    Uses helper function `get_stencil_accessed_symbols` to check which bufs are stencil-accessed in each of these args,
+    and to distinguish the following three cases:
+
+     (1) all accesses in an arg tree are to `buf`  - arg should be moved forward in the computation
+     (2) no accesses are to `buf`                  - arg should be moved backward in the computation
+     (3) there's a mix                             - unknown, take any or no action
+
+    If two args are identified that should be moved forward, or two args are identified that should be moved backwards,
+    the computation is restructured accordingly.
+    """
```

Comment:
> I wonder if a better way to think about it would be a sort-of graph partitioning algorithm, where you want to re-order your graph such that in the end you have as few edges as possible between the two partitions.

It should only be a single edge, to be precise.

> I was thinking that it should be possible to do this with a local rule like:
> 
> match `(a+b)+c` if `source(a) != source(b)` then rewrite to `(same_source_as(a)+a)+not_same_source_as(a))`.

I was hoping that too, but to untangle the structure you'd want to move good things forward *and* bad things backward.

> Also, the terminology of forward/backward seems to be a bit odd, as we are looking at some sort of DAG.

One that is flattened to a list.

> Finally, since all arith ops are pure anyway, the only order that really matters here is use-def order, so I don't really see a point in moving these ops around to be above/below a certain point...

Without the insert point maths, it kept creating scenarios where values are used before they are defined. It seemed best to insert each newly created instruction at a point where both its operands are present to avoid this side-effect of tree restructuring.

Reply:
@PapyChacal suggested to orient this along the lines of what constant folding does.